gh-122450: Expand documentation for ``Rational`` and ``Fraction`` (#136800)
python/cpython: Doc/library/fractions.rst

The :mod:`fractions` module provides support for rational number arithmetic.

> A Fraction instance can be constructed from a pair of numbers, from
> another number, or from a string.

Reviewer: What do you think about this?
```suggestion
A Fraction instance can be constructed from a pair of rational numbers, from
a single number, or from a string.
```

Author: Indeed, this better describes two-arg case.

   The first version requires that *numerator* and *denominator* are instances
   of :class:`numbers.Rational` and returns a new :class:`Fraction` instance
   with a value equal to ``numerator/denominator``.
   The numerator and denominator of this new instance are converted to be in

>    simplest integral terms, always with a positive denominator.

Author: ```suggestion
```

I think above sentence should be removed: we already have descriptions for numerator/denominator attributes.  This will be a *third* time.  Also, notion of denominator clash with denominator *argument name* in surrounding sentences.  Last, but not least: readers can think, that our *internal* representation of the Fraction is like described.  While it's true - this is an implementation detail.

@AA-Turner, please don't merge this.  Lets wait for a second approval from @serhiy-storchaka or @picnixz.

Reviewer 2: > This will be a _third_ time.

In total, yes, but the first and only within the constructor. We have seen from analytics and annecdotal evidence that when looking for reference documentation, readers generally only look at the relevant function/class/method, rather than following links or looking elsewhere. Here, the repetition helps, not hinders.

> Last, but not least: readers can think, that our _internal_ representation of the Fraction is like described. While it's true - this is an implementation detail.

I'm not sure I follow. Could you elaborate?

Reviewer: I concur with @skirpichev. This sentence is redundant here, it will only confuse users.

When you say "The numerator and denominator of this new instance are converted" you refers to the `numerator` and `denominator` attributes. Since there is no any mark up, it can be confused with constructor parameters. If you add links, this will be just a repetition of what was referenced.

The point is not that this sentence can be interpreted in correct way (if you already know how it all works), but that it can be interpreted incorrectly, and it is easy to misinterpret it on first reading.

Reviewer 2: Do you have a suggested rewording? I remain of the opinion that we should mention this in the documentation for the constructor, rather than indirectly through attribute documentation.

Author: > This sentence is redundant here, it will only confuse users.

Does the rest looks ok?

> Since there is no any mark up, it can be confused with constructor parameters.

(As a side note, I don't think it's a good idea if we will distinguish different meanings just by markup.)

> I'm not sure I follow. Could you elaborate?

Rational numbers are just a set of ordered pairs of integers with some equivalency relation.  Any such equivalent pair can be used to represent a rational number.  So, on which numerator/denominator you are talking here?  Are you saying that Fraction internally represented by a canonical representative (in lowest terms and with a positive denominator)?  Or that attributes of the Fraction instance returns that canonical representative?

Reviewer: Why is it needed? In what cases the user needs such information?

Reviewer 2: @skirpichev I'm content for you to remove this sentence (lines 29 & 30).

   The first version requires that *numerator* and *denominator* are instances
   of :class:`numbers.Rational` and returns a new :class:`Fraction` instance

>    with value ``numerator/denominator``. If *denominator* is ``0``, it

Author: ```suggestion
   with value equal to ``numerator/denominator``. If *denominator* is ``0``, it
```
?

Reviewer: I would replace "0" with "zero". It raises not only if the denominator is an integer 0.

>    raises a :exc:`ZeroDivisionError`.

Reviewer 2: I still think this is worse. We now loose information that the denominator is positive, and implicitly require the user to know that `{Fraction,Rational}.denominator` *should* be positive. We are able to guarantee here that the denominator *is* a non-zero positive integer, there's no reason why we can't say that.

How about:

```
... and returns a new :class:`Fraction` instance with a value equal
to ``numerator/denominator``. The numerator and denominator of this
instance will be in the lowest terms, and the denominator will be positive.
```

Author: > We now loose information that the denominator is positive, and implicitly require the user to know that {Fraction,Rational}.denominator should be positive.

No.  I think that class constructor should describe it's construction, not attributes, right?

1. This information will be present in descriptions of numerator/denominator attributes of the Fraction
2. Or in Rational docs.

I would prefer (2) for less redundancy.  But as you insist on restoring (1) - I'll do that.

Author: @AA-Turner, numerator/denominator properties are restored.

>    raises a :exc:`ZeroDivisionError`.

Reviewer 2: How about?:
```suggestion
   with a value equal to ``numerator/denominator``.
   The numerator and denominator of this new instance are converted to be in
   simplest integral terms, always with a positive denominator.
   If *denominator* is zero, it raises a :exc:`ZeroDivisionError`.
```

Author: I don't think we should more test readers intelligence.  We have description of numerator and denominator attributes.  Now twice: in the Rational docs and in the Fraction's docs.

I'm fine with wording of your suggestion, just think that this (third!) description of numerator/denominator properties is overkill.  Feel free to apply this, however.

      The :class:`Fraction` constructor now accepts any objects with the
      :meth:`!as_integer_ratio` method.

>    .. attribute:: numerator

Reviewer: On other hand, `Fraction` provides additional guarantees -- "in lowest terms", and the denominator is positive. So why not leave this?

Author: > Fraction provides additional guarantees -- "in lowest terms", and the denominator is positive.

No!  It's required already for number.Rational.

N/B: I'm fine with leaving this (per Adam suggestion in other pr), but it seems redundant a bit.

Reviewer: I see. It is in the paragraph above description of `numerator` and `denominator`.

I do not see any mention of the relation between `Fraction` and `Rational`. We should say that `Fraction` subtypes `Rational` and also provides such and such methods. Since concrete implementation of `__floor__()` etc is documented (they are abstract in `Real`), we should also document concrete implementations of `numerator` and `denominator`. We can add some examples, e.g `Fraction(15, -20)` -> `-3, 4`.

Author: > I see. It is in the paragraph above description of numerator and denominator.

docstrings for above properties are more verbose.

> I do not see any mention of the relation between Fraction and Rational.

It's there, just after examples.  I think we should move this up, to the first paragraph.  Does make sense?

> Since concrete implementation of ``__floor__()`` etc is documented (they are abstract in Real), we should also document concrete implementations of numerator and denominator.

But this not adds something c.f. Rational docs.  All we can say is, again: numerator/denominator are in lowest terms and denominator>0.

I.e. we can't also say, that these are int's:
```pycon
>>> from gmpy2 import mpz
>>> from fractions import Fraction
>>> Fraction(mpz(1), mpz(2))
Fraction(1, 2)
>>> a = _
>>> a.numerator
mpz(1)
>>> a.denominator
mpz(2)
```

BTW, all dunders methods in the Fraction sphinx docs looks redundant, except for the ``__format__()`` description.

> We can add some examples, e.g Fraction(15, -20) -> -3, 4.

We have such example:
```pycon
>>> Fraction(16, -10)
Fraction(-8, 5)
```

Reviewer 2: @skirpichev please keep this. We shouldn't force people to read documentation for the numeric tower when all they need is a quick reference for `Fraction`. It is far more pragmatic to keep the text.

   .. attribute:: denominator

>       Denominator of the Fraction in lowest term.  Should be positive.

Reviewer 2: ```suggestion
      Denominator of the Fraction in lowest terms.
      Guaranteed to be positive.
```
